node/src/common/AsyncQueue.h

```diff
@@ -0,0 +1,252 @@
+//
+// Copyright (c) 2015, Microsoft Corporation
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
+// IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+//
+
+namespace OpenT2T
+{
+
+// Interface implemented by callers
+template <class QueueItem>
+class IQueueItemHandler
+{
+public:
+    virtual void OnStarted() = 0;
+    virtual void OnProcessQueueItem(QueueItem& item) = 0;
+    virtual void OnStopped() = 0;
+};
+
+static bool s_crtIsTerminating = false;
+static bool s_registered_atexit_handler = false;
+
+// Implements a FIFO queue that has its own worker thread.
+// Callers must register a handler interface during initialization.
+// Callers can push items onto the queue. The registered interface is then notified
+// from the AsyncQueue's worker thread to process each queued item.
+template <class QueueItem>
+class AsyncQueue
+{
+public:
+    AsyncQueue() :
+        _stopWorkerThread(false),
+        _workerThreadStopped(false),
+        _handler(nullptr),
+        _isInitialized(false)
+    {
+        // Ensure that we register an "atexit" handler for the CRT.
+        // This handler is used to let us know when the CRT is terminating.
+        if (!s_registered_atexit_handler)
+        {
+            if (std::atexit(AsyncQueue_atexit_handler) == 0)
+            {
+                s_registered_atexit_handler = true;
+            }
+        }
+    };
+
+    ~AsyncQueue()
+    {
+        Uninitialize();
+    }
+
+    void Initialize(const std::shared_ptr<IQueueItemHandler<QueueItem>>& handler)
+    {
+        std::lock_guard<std::mutex> lock(_mutex);
+
+        if (handler == nullptr) throw new std::invalid_argument("handler cannot be null");
+
+        // Verify that this async queue has not already been initialized
+        if (!_isInitialized)
```

Comment:
I didn't write this code... but I don't think a local function static variable would be a good solution here because there could be more than one instance of this class.

---
In reply to: [71086953](https://github.com/openT2T/client/pull/1#discussion_r71086953) [](ancestors = 71086953)